blivet/iscsi.py

```diff
@@ -277,10 +319,41 @@ def discover(self, ipaddr, port="3260", username=None, password=None,
                                                  reverse_password=r_password)
             self.startup()
 
-            # Note may raise an IOError
-            found_nodes = libiscsi.discover_sendtargets(address=ipaddr,
-                                                        port=int(port),
-                                                        authinfo=authinfo)
+            # start libiscsi discover_sendtargets in a new process
+            # threads can't be used here because the libiscsi library
+            # using signals internally which are send to bad thread
+            (con_recv, con_write) = Pipe(False)
+            p = Process(target=_call_discover_targets, args=(con_write,
+                                                             ipaddr,
+                                                             port,
+                                                             authinfo, ))
+
+            p.start()
+
+            found_nodes = []
+
+            try:
+                (ok, data) = con_recv.recv()
+
+                if not ok:
+                    found_nodes =  None
+                    log.debug("iSCSI: error raised when "
+                              "discover_sendtargets process called: %s",
+                              str(data))
+
+            except EOFError:
+                found_nodes = None
+                log.debug("iSCSI: can't receive response from "
+                          "_call_discover_targets")
+
+            p.join()
+
+            # convert dictionary back to iscsi nodes object
+            if found_nodes is not None:
+                for node in data:
+                    found_nodes.append(libiscsi.node(**node))
+
+
```

Comment:
Looks good to me. The only minor concern for me is the amount of empty space in the new code. I think you can get rid of some blank lines -- before ``p.start()``, before ``if not ok``, before ``except EOFError`` and at the very end.